Add map tool to edit blank segments

troopa81 opened the pull request:
This is the map tool to edit blank segments described in QEP [#345](https://github.com/qgis/QGIS-Enhancement-Proposals/pull/345)

![blanksegments_maptool](https://github.com/user-attachments/assets/4b357590-ec8f-4d93-a26f-8d70213c6c40)

It allows to define blank segments per feature on templated line symbology. Hashes or Markers won't be drawn inside those blank segments. The blank segments are stored on a data defined property (data field or auxiliary storage field). This tool allows to:
- Click in the neighborhood of the line to start creating a segment, click again to finish editing
- Select a segment, press Del key to remove it
- Drag the start/end of an already existing segment and move it along the line to reduce/enlarge it

**Funded by Stadt Frankfurt am Main and Oslandia**

troopa81 committed:
feat(BlankSegments): Add maptool to edit blank segments

troopa81 committed:
fix(BlankSegments): fix doc and moc

nyalldawson commented:
I haven't reviewed the code, but the UI needs some love here:

- remove the unnecessary left padding
- the icon should follow the conventions used by the "show/hide labeling tool". Ie "eye" motif instead of glasses, removing the red "delete" overlay
<img width="58" height="53" alt="image" src="https://github.com/user-attachments/assets/55e5b4c3-4210-4a68-98f6-8caeeb29b68b" />
- add explanatory tooltip
- For logical flow I think the button should be placed last in that horizontal list, not first. (@nirvn what do you think?)
- Blank segments group box label needs to follow HIG for capitalization. This should be a collapsible group box, collapsed by default to avoid UI clutter

troopa81 committed:
fix(BlankSegments): Update tooltip according to symbol type

troopa81 committed:
ui(BlankSegments): Remove left padding and change icon

troopa81 committed:
ui(BlankSegments): collapse by default

troopa81 committed:
ui(BlankSegments): Use QgsCollapsibleGroupBox

and not QgsCollapsibleGroupBoxBasic, so state is stored/retrieve
within the settings

troopa81 committed:
style(BlankSegments): Silence clang-tidy false alarm

troopa81 committed:
tests(BlankSegments): Remove broken tests

That requires code to make it work

benoitdm-oslandia commented on src/core/symbology/line_p.h:
```diff
@@ -0,0 +1,92 @@
+
+///@cond PRIVATE
+
+class Line
```
Can you put some comment please? What is this class for?

benoitdm-oslandia commented on src/core/symbology/line_p.h:
```diff
@@ -0,0 +1,92 @@
+    Line( QPointF p1, QPointF p2 )
+      : mVertical( false )
+      , mIncreasing( false )
+      , mT( 0.0 )
```
please avoid too short var name

benoitdm-oslandia commented on src/gui/maptools/qgsmaptooleditblanksegments.cpp:
```diff
@@ -0,0 +1,947 @@
+namespace
+{
+
+  class SymbolLayerVisitor : public QgsStyleEntityVisitorInterface
```
please add comment just to explain the main goal/use of this class

benoitdm-oslandia commented on src/gui/maptools/qgsmaptooleditblanksegments.cpp:
```diff
@@ -0,0 +1,947 @@
+          if ( sl->id() == mSymbolLayerId )
+          {
+            mSymbolLayer = dynamic_cast<const QgsTemplatedLineSymbolLayerBase *>( sl );
+            mSymbolLayerIndex = idx;
```
what if the cast failed? you will have a null mSymbolLayer but a defined mSymbolLayerIndex?

benoitdm-oslandia commented on src/gui/maptools/qgsmaptooleditblanksegments.cpp:
```diff
@@ -0,0 +1,947 @@
+    if ( visitor.symbol() && visitor.symbolLayer() && visitor.symbolLayerIndex() > -1 )
+    {
+      mSymbol.reset( visitor.symbol()->clone() );
+      if ( mSymbolLayer = createRenderedPointsSymbolLayer( visitor.symbolLayer() ); mSymbolLayer )
```
ok, this is official: I am not fond of this `if` usage! But this is only a aesthetic remark, I do not know if this is common or *accepted* elsewhere in QGIS code. Furthermore, I think you can just do: 

```suggestion
      if ( mSymbolLayer = createRenderedPointsSymbolLayer( visitor.symbolLayer() ) )
```

benoitdm-oslandia commented on src/gui/maptools/qgsmaptooleditblanksegments.cpp:
```diff
@@ -0,0 +1,947 @@
+}
+
+
+const QPointF &pointAt( const QList<QList<QPolygonF>> &points, int partIndex, int ringIndex, int pointIndex )
```
can you move it at the top in the nonamed namespace?

benoitdm-oslandia reviewed (commented):
This is a big PR I will need to build it and test it but it looks nice! The feature is complex to set in place so is the code. The code is a bit tricky and needs more comments (more) and maybe the code files can be splat to be more readable.

I made many nitpicking comments for this review but I will make a more thoughtful review next week!

troopa81 committed:
doc(BlankSegments): Add some comments

troopa81 reviewed (commented):
@benoitdm-oslandia I treated all the comments (except the ones I didn't fully agree with)

I had to rebase on master because of moved files/code

troopa81 committed:
style(BlankSegments): rename BlankSegment rubber band

And move it to cpp

troopa81 committed:
doc(BlankSegments): Add methods comments

And move cpp only method to anonymous namespace

troopa81 committed:
style(BlankSegments): factorize units into a const variable

troopa81 commented on src/gui/symbology/qgssymbollayerwidget.cpp:
```diff
@@ -1890,10 +1892,13 @@ QgsTemplatedLineSymbolLayerWidget::QgsTemplatedLineSymbolLayerWidget( TemplatedS
   mOffsetAlongLineUnitWidget->setUnits( QgsUnitTypes::RenderUnitList() << Qgis::RenderUnit::Millimeters << Qgis::RenderUnit::MetersInMapUnits << Qgis::RenderUnit::MapUnits << Qgis::RenderUnit::Pixels << Qgis::RenderUnit::Points << Qgis::RenderUnit::Inches << Qgis::RenderUnit::Percentage );
   mAverageAngleUnit->setUnits( QgsUnitTypes::RenderUnitList() << Qgis::RenderUnit::Millimeters << Qgis::RenderUnit::MetersInMapUnits << Qgis::RenderUnit::MapUnits << Qgis::RenderUnit::Pixels << Qgis::RenderUnit::Points << Qgis::RenderUnit::Inches );
+  mIntervalUnitWidget->setUnits( QgsUnitTypes::RenderUnitList() << Qgis::RenderUnit::Millimeters << Qgis::RenderUnit::MetersInMapUnits << Qgis::RenderUnit::MapUnits << Qgis::RenderUnit::Pixels << Qgis::RenderUnit::Points << Qgis::RenderUnit::Inches );
+  mBlankSegmentsUnitWidget->setUnits( QgsUnitTypes::RenderUnitList() << Qgis::RenderUnit::Millimeters << Qgis::RenderUnit::MetersInMapUnits << Qgis::RenderUnit::MapUnits << Qgis::RenderUnit::Pixels << Qgis::RenderUnit::Points << Qgis::RenderUnit::Inches );
```
I used a const variable to factorize it

troopa81 committed:
fix(BlankSegments): fix spell check and documentation

troopa81 committed:
fix(BlankSegments): fix CI

troopa81 committed:
fix(BlankSegments): fix documentation

troopa81 committed:
fix(BlankSegments): fix documentation (again)

troopa81 commented:
unrelated mac os build failure

nyalldawson commented:
@troopa81 can you please re-familiarise yourself with the coding guidelines? https://github.com/qgis/QGIS-Enhancement-Proposals/blob/master/qep-314-coding-style.md

I was looking to fix the warnings this PR has introduced, and noticed that the naming conventions used throughout SymbolLayerFinder/ BlankSegmentRubberBand/ QgsMapToolEditBlankSegmentsBase violiate the guidelines.

troopa81 commented:
> I'm sorry you feel that way, but I don't have the time or motivation to individually correct things like this.

I don't remember asking you correcting anything. If there are things to be fixed, I'll do it.

> As a frequent core contributor you should already be aware of the coding style guidelines and follow them without needing someone to point this out in review.

I believed that what the review process was for, detecting developer mistakes. I do everything I can to comply to the written and unwritten rules of the project, and sometimes I make mistakes. I don't think it deserves the type of harsh comments I read here.
